ta/remoteproc/remoteproc_core.c

```diff
@@ -0,0 +1,1098 @@
+	 * getting the key object from secure storage.
+	 */
+
+	/* By default ask the remoteproc PTA to verify the signature. */
+	res = remoteproc_pta_verify(ctx, algo, hash, hash_len);
+
+out:
+	TEE_FreeOperation(op);
+free_hash:
+	TEE_Free(hash);
+
+	return res;
+}
+
+static TEE_Result remoteproc_verify_header(struct remoteproc_context *ctx,
+					   uint32_t fw_orig_size)
+{
+	struct remoteproc_fw_hdr *hdr = (void *)ctx->sec_cpy;
+	uint32_t size = 0;
+
+	if (hdr->magic != RPROC_HDR_MAGIC)
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	if (hdr->version != HEADER_VERSION)
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	/*
+	 * The offsets are aligned to 64 bits format. While the length of each
+	 * chunks are the effective length, excluding the alignment padding
+	 * bytes.
+	 */
+	if (ADD_OVERFLOW(sizeof(*hdr), ROUNDUP_64(hdr->sign_len), &size) ||
+	    ADD_OVERFLOW(size, ROUNDUP_64(hdr->img_len), &size) ||
+	    ADD_OVERFLOW(size, ROUNDUP_64(hdr->tlv_len), &size) ||
+	    fw_orig_size != size)
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	return TEE_SUCCESS;
+}
+
+static TEE_Result get_rproc_pta_capabilities(struct remoteproc_context *ctx)
+{
+	uint32_t param_types = TEE_PARAM_TYPES(TEE_PARAM_TYPE_VALUE_INPUT,
+					       TEE_PARAM_TYPE_VALUE_OUTPUT,
+					       TEE_PARAM_TYPE_VALUE_OUTPUT,
+					       TEE_PARAM_TYPE_NONE);
+	TEE_Param params[TEE_NUM_PARAMS] = { };
+	TEE_Result res = TEE_ERROR_GENERIC;
+
+	params[0].value.a = ctx->rproc_id;
+
+	res = TEE_InvokeTACommand(pta_session, TEE_TIMEOUT_INFINITE,
+				  PTA_RPROC_HW_CAPABILITIES,
+				  param_types, params, NULL);
+	if (res)
+		return res;
+
+	ctx->hw_fmt = params[1].value.a;
+	ctx->hw_img_prot = params[2].value.a;
+
+	return TEE_SUCCESS;
+}
+
+static TEE_Result remoteproc_verify_firmware(struct remoteproc_context *ctx,
+					     uint8_t *fw_orig,
+					     uint32_t fw_orig_size)
+{
+	struct remoteproc_fw_hdr *hdr = NULL;
+	TEE_Result res = TEE_ERROR_GENERIC;
+
+	res = get_rproc_pta_capabilities(ctx);
+	if (res)
+		return res;
+
+	/* Secure the firmware image depending on strategy */
+	if (!(ctx->hw_img_prot & PTA_RPROC_HWCAP_PROT_HASH_TABLE) ||
+	    ctx->hw_fmt != PTA_RPROC_HWCAP_FMT_ELF) {
+		/*
+		 * Only hash table for ELF format support implemented
+		 * in a first step.
+		 */
+		return TEE_ERROR_NOT_IMPLEMENTED;
+	}
+
+	res = remoteproc_save_fw_header_and_tlvs(ctx, fw_orig, fw_orig_size);
+	if (res)
+		return res;
+
+	res = remoteproc_verify_header(ctx, fw_orig_size);
+	if (res)
+		goto free_sec_cpy;
+
+	hdr = (void *)ctx->sec_cpy;
+	ctx->tlvs_sz = hdr->tlv_len;
+	ctx->tlvs = FW_TLV_PTR(ctx->sec_cpy, hdr);
+
+	res = remoteproc_verify_signature(ctx);
+	if (res)
+		goto free_sec_cpy;
+
+	/* Store location of the loadable binary in non-secure memory */
+	ctx->fw_img_sz = hdr->img_len;
+	ctx->fw_img = FW_IMG_PTR(fw_orig, hdr);
+
+	DMSG("Firmware images addr: %p size: %zu", ctx->fw_img,
+	     ctx->fw_img_sz);
+
+	return TEE_SUCCESS;
+
+free_sec_cpy:
+	TEE_Free(ctx->sec_cpy);
```

Comment:
I see that `ctx->sec_cpy` is used only in the scope of this function `remoteproc_verify_firmware()`.
I think the field could be removed from `struct remoteproc_context` and be referenced from `remoteproc_verify_firmware()` stack, passed as argument to `remoteproc_save_fw_header_and_tlvs()`, `remoteproc_verify_header()` and `remoteproc_verify_signature()`.

Reply:
The secure copy contains help hash table and TLV. I implemented in this way to be able to reuse it later out of this function.
As today I have no code that reuses it ( ST downstream) I can change this as you propose.
we can come back to a context pointer later if needed.

Reply:
So here if I well understood, I keep the pointer but set it to NULL before free it.
I'm not sureto  understand your point concerning `assert(ctx->sec_cpy)` as function that use the pointer are called by `remoteproc_load_fw()`.
It seems  to me some over checking. But perhaps you prefer that add it anyway?

Reply:
At least, in your latest proposal, `ctx->sec_cpy` is cleared when the buffer is freed. So fine with me.